<a href="https://colab.research.google.com/github/lynseybwisa/IBM-Machine-Learning/blob/develop/Airlines.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Feature Engineering**

Feature engineering involves three steps: 
1. feature transformation (transforming the original features), 
2. feature selection (selecting the most useful features to train on)
3. feature extraction (combining existing features to produce more useful ones). 

## **Installing Required Libraries**

In [21]:
!pip install -qy pandas==1.3.4 numpy==1.21.4 seaborn==0.9.0 matplotlib==3.5.0 scikit-learn==0.20.1


Usage:   
  pip3 install [options] <requirement specifier> [package-index-options] ...
  pip3 install [options] -r <requirements file> [package-index-options] ...
  pip3 install [options] [-e] <vcs project url> ...
  pip3 install [options] [-e] <local project path> ...
  pip3 install [options] <archive url/path> ...

no such option: -y


In [22]:
!pip install -qy openpyxl


Usage:   
  pip3 install [options] <requirement specifier> [package-index-options] ...
  pip3 install [options] -r <requirements file> [package-index-options] ...
  pip3 install [options] [-e] <vcs project url> ...
  pip3 install [options] [-e] <local project path> ...
  pip3 install [options] <archive url/path> ...

no such option: -y


In [23]:
# Surpress warnings from using older version of sklearn:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn

In [24]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import plotly.express as px

from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

## **Data Cleaning**

This dataset contains the prices of flight tickets for various airlines between the months of March and June of 2019 and between various cities. This dataset is often used for prediction analysis of the flight prices which are influenced by various factors, such as name of the airline, date of journey, route, departure and arrival times, the source and the destination of the trip, duration and other parameters.

Read the data into pandas data frame and looking at the first 5 rows using the head() method.

In [25]:
data = pd.read_excel('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-ML0232EN-SkillsNetwork/asset/airlines_data.xlsx')
data.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

By using the info function, take a look at the types of data that our dataset contains.

In [26]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


total rows = 10683, total columns = 11, most columns are object data types, price is integer

The describe() function provides the statistical information about the numerical variables. In our case, it is the 'price' variable.



In [27]:
data.describe()

Price
count  10683.000000
mean    9087.064121
std     4611.359167
min     1759.000000
25%     5277.000000
50%     8372.000000
75%    12373.000000
max    79512.000000

Next, check for any null values.This gives a count of the number of missing values in each column of the DataFrame.

Specifically, the isnull() method will return a DataFrame of the same shape as the input DataFrame, where each value is either True if it is missing, or False if it is not missing. Then, the sum() method is called on this DataFrame to count the number of True values in each column, which gives you the number of missing values in each column.

In [28]:
data.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

Now that we have found some null points, we need to either remove them from our dataset or fill them with something else. In this case, we will use fillna() and method='ffill', which fills the last observed non-null value forward until another non-null value is encountered.

In [29]:
data = data.fillna(method='ffill')

## **Feature Transformation**

Feature Transformation means transforming our features to the functions of the original features. For example, feature encoding, scaling, and discretization (the process of transforming continuous variables into discrete form, by creating bins or intervals) are the most common forms of data transformation.

### **Dealing with Categorical Variables**

Categorical variables represent qualitative data with no apparent inherent mathematical meaning. Therefore, for any machine learning analysis, all the categorical data must be transformed into the numerical data types. First, we'll start with 'Airlines' column, as it contains categorical values. We will use `unique()` method to obtain all the categories in this column.


In [30]:
data['Airline'].unique().tolist()

['IndiGo',
 'Air India',
 'Jet Airways',
 'SpiceJet',
 'Multiple carriers',
 'GoAir',
 'Vistara',
 'Air Asia',
 'Vistara Premium economy',
 'Jet Airways Business',
 'Multiple carriers Premium economy',
 'Trujet']

From the above list, we notice that some of the airline names are being repeated. For example, 'Jet Airways' and 'Jet Airways Business'. This means that some of the airlines are subdivided into separate parts. We will combine these 'two-parts' airlines to make our categorical features more consistent with the rest of the variables.

Here, we will use the *numpy* `where()` function to locate and combine the two categories.

In [31]:
data['Airline'] = np.where(data['Airline']=='Vistara Premium economy', 'Vistara', data['Airline'])
data['Airline'] = np.where(data['Airline']=='Jet Airways Business', 'Jet Airways', data['Airline'])

The np.where() function is a NumPy function that takes three arguments: a condition, a value to replace the condition with if it is True, and a value to use if the condition is False. In this case, the condition is whether the 'Airline' column of data is equal to a certain string (either 'Vistara Premium economy' or 'Jet Airways Business'), and the values to use if the condition is True are 'Vistara' and 'Jet Airways', respectively.

By using np.where() in this way, we are essentially performing a find-and-replace operation on the 'Airline' column of data

combine 'Multiple carriers Premium economy' and 'Multiple carriers' categories. Print the newly created list using `unique().tolist()` functions.

In [32]:
data['Airline'] = np.where(data['Airline']=='Multiple carriers Premium economy', 'Multiple carriers', data['Airline'])

In [33]:
data['Airline'].unique().tolist()

['IndiGo',
 'Air India',
 'Jet Airways',
 'SpiceJet',
 'Multiple carriers',
 'GoAir',
 'Vistara',
 'Air Asia',
 'Trujet']

#### **One Hot Encoding**

We will use, `get_dummies()` method to do this transformation. In the next cell, we will transform 'Airline', 'Source', and 'Destination' into their respective numeric variables. We will put all the transformed data into a 'data1' data frame.

In [34]:
data1 = pd.get_dummies(data=data, columns = ['Airline', 'Source', 'Destination'])

Tthe get_dummies() function will create new binary columns for each category in the three variables. For example, if the Airline column has categories IndiGo, Air India, and Jet Airways, then the get_dummies() function will create three new columns: Airline_IndiGo, Airline_Air India, and Airline_Jet Airways. Each of these columns will contain binary values indicating whether the observation belongs to that category (1) or not (0).

In [35]:
data1.head()

Date_of_Journey                  Route Dep_Time  Arrival_Time Duration  \
0      24/03/2019              BLR → DEL    22:20  01:10 22 Mar   2h 50m   
1       1/05/2019  CCU → IXR → BBI → BLR    05:50         13:15   7h 25m   
2       9/06/2019  DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h   
3      12/05/2019        CCU → NAG → BLR    18:05         23:30   5h 25m   
4      01/03/2019        BLR → NAG → DEL    16:50         21:35   4h 45m   

  Total_Stops Additional_Info  Price  Airline_Air Asia  Airline_Air India  \
0    non-stop         No info   3897                 0                  0   
1     2 stops         No info   7662                 0                  1   
2     2 stops         No info  13882                 0                  0   
3      1 stop         No info   6218                 0                  0   
4      1 stop         No info  13302                 0                  0   

   ...  Source_Chennai  Source_Delhi  Source_Kolkata  Source_Mumbai  \
0  ...               0             0               0              0   
1  ...               0             0               1              0   
2  ...               0             1               0              0   
3  ...               0             0               1              0   
4  ...               0             0               0              0   

   Destination_Banglore  Destination_Cochin  Destination_Delhi  \
0                     0                   0                  0   
1                     1                   0                  0   
2                     0                   1                  0   
3                     1                   0                  0   
4                     0                   0                  0   

   Destination_Hyderabad  Destination_Kolkata  Destination_New Delhi  
0                      0                    0                      1  
1                      0                    0                      0  
2                      0                    0                      0  
3                      0                    0                      0  
4                      0                    0                      1  

[5 rows x 28 columns]

compare our original data frame with the transformed one.

In [36]:
data.shape

(10683, 11)

In [37]:
data1.shape

(10683, 28)

use `value_counts()` to determine the values distribution of the 'Total_Stops' parameter.

In [40]:
value_counts = data['Total_Stops'].value_counts()
value_counts

1 stop      5625
non-stop    3492
2 stops     1520
3 stops       45
4 stops        1
Name: Total_Stops, dtype: int64

#### **Label Encoding**